### PR TITLE
Remove strings from typed_absy

### PR DESCRIPTION
**Pull request by Schaeff**
- removes strings in typed_absy
- adapt semantic check to keep working (less use of references, cloning)
- adapt flattening to work with non-string-based identifiers, especially for arrays as we used to use "foo_c42"  for the 43th item of `foo`. Now we instead keep track of all flat variables for a given typed variable and access the 43th item of that.

### CODE REVIEW

File: zokrates_core/src/flatten/mod.rs [hunk 1]
```diff
@@ -11,19 +11,18 @@ use crate::typed_absy::*;
...
 
 /// Flattener, computes flattened program.
 #[derive(Debug)]
-pub struct Flattener {
+pub struct Flattener<'ast> {
     /// Index of the next introduced variable while processing the program.
     next_var_idx: usize,
     ///
-    bijection: BiMap<String, FlatVariable>,
+    layout: HashMap<Identifier<'ast>, Vec<FlatVariable>>,
```
Comment (Schaeff) [inline zokrates_core/src/flatten/mod.rs]: Any Identifier in the TypedProgram maps to a series of FlatVariable (one for Field, n for FieldArray(n))

File: zokrates_core/src/flatten/mod.rs [hunk 2]
```diff
@@ -725,8 +722,8 @@ impl Flattener {
                         FieldElementArrayExpression::Identifier(size, id) => {
                             assert!(n < T::from(size));
                             FlatExpression::Identifier(
-                                self.get_latest_var_substitution(&format!("{}_c{}", id, n))
-                                    .clone(),
+                                self.layout.get(&id).unwrap().clone()
+                                    [n.to_dec_string().parse::<usize>().unwrap()],
```
Comment (Schaeff) [inline zokrates_core/src/flatten/mod.rs]: get flat variables for id, take the nth. Could be made cleaner

File: zokrates_core/src/flatten/mod.rs [hunk 3]
```diff
@@ -1177,35 +1185,12 @@ impl Flattener {
...
-                                }
-                            }
-                        }
+                        let rhs = rhs_flattened.expressions.into_iter();
+
+                        let vars = vars.into_iter().flat_map(|v| self.use_variable(&v));
 
-                        // we should have exhausted the return values
-                        assert_eq!(iterator.next(), None);
+                        statements_flattened
```
Comment (Schaeff) [inline zokrates_core/src/flatten/mod.rs]: use `zip` instead of low level iterator magic

File: zokrates_core/src/flatten/mod.rs [hunk 4]
```diff
@@ -1299,30 +1265,31 @@ impl Flattener {
     /// # Arguments
     ///
     /// * `name` - a String that holds the name of the variable
-    fn use_variable(&mut self, name: &String) -> FlatVariable {
-        // issue the variable we'll use
-        let var = self.issue_new_variable();
+    fn use_variable(&mut self, variable: &Variable<'ast>) -> Vec<FlatVariable> {
```
Comment (Schaeff) [inline zokrates_core/src/flatten/mod.rs]: We issue flat variables for typed variables, whatever the type

File: zokrates_core/src/static_analysis/inline.rs [hunk 5]
```diff
@@ -5,14 +5,14 @@ use crate::types::{Signature, Type};
...
 use zokrates_field::field::Field;
 
-pub struct Inliner<T: Field> {
-    functions: Vec<TypedFunction<T>>,
-    statements_buffer: Vec<TypedStatement<T>>,
-    context: Vec<(String, usize)>,
+pub struct Inliner<'ast, T: Field> {
+    functions: Vec<TypedFunction<'ast, T>>,
+    statements_buffer: Vec<TypedStatement<'ast, T>>,
+    context: Vec<(&'ast str, Signature, usize)>,
```
Comment (Schaeff) [inline zokrates_core/src/static_analysis/inline.rs]: the context is function name + signature + number of call

File: zokrates_core/src/static_analysis/mod.rs [hunk 6]
```diff
@@ -33,6 +33,7 @@ impl<T: Field> Analyse for TypedProg<T> {
         let r = Propagator::propagate(r);
         // apply inlining strategy
         let r = Inliner::inline(r);
+        println!("{}", r);
```
Comment (JacobEberhardt) [inline zokrates_core/src/static_analysis/mod.rs]: should this be here?

**Pull request merged**